app/controllers/badges_controller.rb

```diff
@@ -3,7 +3,7 @@ class BadgesController < ApplicationController
   # No authorization required for entirely public controller
 
   def show
-    @badge = Badge.find_by_slug(params[:slug])
+    @badge = Badge.find_by_slug(params[:slug]) || not_found
```

Comment:
I suggest `@badge = Badge.find_by!(slug: params[:slug])` . Using the bang method will raise `ActiveRecord::RecordNotFound` which makes more sense. The exception will be caught by rails 404 page will be rendered.

Reply:
Even better, I was trying to follow the style I found around the code, though improving on it it's definitely a plus 💃

Reply:
I feel like I was just doing it the way I saw someone else do it somewhere randomly online a few years ago and never really second-guessed the approach. Then it just became a pattern.

The code is filled with stuff like that, but we're digging our way out.

I like @lightalloy's suggestion but I also don't mind the current approach, which makes it super clear what's going on.

Reply:
I do think @lightalloy's solution is cleaner, but have a couple of things to point out here in favour of explicit `not_found`

**Advantage**
- It makes it obvious that the person who wrote the code has thought about `404` and have handled this scenario.
- It's not obvious that the feature **requires** the `!` -- someone trying to build a new feature might see this and thought there's no need for the bang, ensures a next round of the same bug

I'm still okay to change if the consensus is to use bang, what do you guys think?

Reply:
I agree that the current approach is more clear :thinking:
Since we're discussing it, I'll add another thought on `not_foun` optimization if we keep the current approach. Currently, we're raising `ActionController::RoutingError` in the `not_found` method, but that's kind of using exceptions as flow control, which is slow. That's not actually an exceptional situation and we don't notify Airbrake of it. So the `not_found` method could just render the corresponding template and return 404 status, like:
```ruby
render file: "#{Rails.root}/public/404.html", layout: false, status: :not_found
```
@ben @rhymes What do you think about this?

Reply:
@lightalloy if you rewrite 

```ruby
  def not_found
    raise ActionController::RoutingError, "Not Found"
  end
```

to become

```
  def not_found
    render file: "#{Rails.root}/public/404.html", layout: false, status: :not_found
  end
```

doesn't it mean that, in the various instances where it's used for example:

https://github.com/thepracticaldev/dev.to/blob/f021a1450140dd63554bc50d480cb1b6a4745acd/app/controllers/badges_controller.rb#L5-L8

the control will go back to the calling function and it will execute everything after the render? In this case we don't want to execute anything after the `not_found` because it might trigger side effects.

In this example

```ruby
def show
  render file: "#{Rails.root}/public/404.html", layout: false, status: :not_found
  puts "i'm here after the render"
end
```
the line after render is executed.

I think that's why Rails by default uses the exception as an escape hatch, to avoid this situation.

Reply:
(you could add a `return not_found` but then you're back at the problem of making sure that someone doesn't forget the return or you'll have unintented consequences in the code)

My vote goes to

```ruby
def not_found
  raise ActiveRecord::RecordNotFound, "Not Found"
end
```

which is what the `!` raise anyway, this way we make it explicit but don't deviate from Rails default

Reply:
@rhymes Yep, I'm familiar to this problem. The change I suggested would require changing the code, where the method is used, which is probably not optimal.
I like your suggestion to alter the `not_found` method, this way the intentions are more clear. But probably it should be a topic of another pr.

Reply:
I'll do it in a separate PR once this is merged, if that's okay with you